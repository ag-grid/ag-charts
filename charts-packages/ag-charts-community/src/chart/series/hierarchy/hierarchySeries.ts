import { HierarchyChart } from "../../hierarchyChart";
import { Series } from "../series";
import { reactive } from "../../../util/observable";

export abstract class HierarchySeries extends Series {
    chart?: HierarchyChart;

    @reactive('dataChange') data: any = undefined;
}