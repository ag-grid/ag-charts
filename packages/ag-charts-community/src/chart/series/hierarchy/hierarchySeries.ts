import type { Point } from '../../../integrated-charts-scene';
import type { ModuleContext } from '../../../module/moduleContext';
import { ColorScale } from '../../../scale/colorScale';
import type { PointLabelDatum } from '../../../util/labelPlacement';
import { OPT_COLOR_STRING_ARRAY, OPT_STRING, Validate } from '../../../util/validation';
import { DEFAULT_FILLS, DEFAULT_STROKES } from '../../themes/defaultColors';
import { Series, SeriesNodePickMode } from '../series';
import type { ISeries, SeriesNodeDatum } from '../seriesTypes';

type Mutable<T> = {
    -readonly [k in keyof T]: T[k];
};

export class HierarchyNode implements SeriesNodeDatum {
    static Walk = {
        PreOrder: 0,
        PostOrder: 1,
    };

    readonly midPoint: Point;

    constructor(
        public readonly series: ISeries<any>,
        public readonly index: number,
        public readonly datum: Record<string, any> | undefined,
        public readonly size: number,
        public readonly fill: string | undefined,
        public readonly stroke: string | undefined,
        public readonly sumSize: number,
        public readonly depth: number | undefined,
        public readonly parent: HierarchyNode | undefined,
        public readonly children: HierarchyNode[]
    ) {
        this.midPoint = { x: 0, y: 0 };
    }

    contains(other: HierarchyNode): boolean {
        let current: HierarchyNode | undefined = other;
        // Index check is a performance optimization - it does not affect correctness
        while (current != null && current.index >= this.index) {
            if (current === this) {
                return true;
            }
            current = current.parent;
        }
        return false;
    }

    walk(callback: (node: HierarchyNode) => void, order = HierarchyNode.Walk.PreOrder) {
        if (order === HierarchyNode.Walk.PreOrder) {
            callback(this);
        }

        this.children.forEach((child) => {
            child.walk(callback, order);
        });

        if (order === HierarchyNode.Walk.PostOrder) {
            callback(this);
        }
    }

    *[Symbol.iterator](): Iterator<HierarchyNode> {
        yield this;

        for (const child of this.children) {
            yield* child;
        }
    }
}

export abstract class HierarchySeries<S extends SeriesNodeDatum> extends Series<S> {
    @Validate(OPT_STRING)
    childrenKey?: string = 'children';

    @Validate(OPT_STRING)
    sizeKey?: string = undefined;

    @Validate(OPT_STRING)
    colorKey?: string = undefined;

    @Validate(OPT_COLOR_STRING_ARRAY)
    fills: string[] = Object.values(DEFAULT_FILLS);

    @Validate(OPT_COLOR_STRING_ARRAY)
    strokes: string[] = Object.values(DEFAULT_STROKES);

    @Validate(OPT_COLOR_STRING_ARRAY)
    colorRange?: string[] = undefined;

    rootNode = new HierarchyNode(this, 0, undefined, 0, undefined, undefined, 0, undefined, undefined, []);
    maxDepth = 0;

    constructor(moduleCtx: ModuleContext) {
        super({
            moduleCtx,
            pickModes: [SeriesNodePickMode.EXACT_SHAPE_MATCH],
            contentGroupVirtual: false,
        });
    }

    getLabelData(): PointLabelDatum[] {
        return [];
    }

    override hasData() {
        return Array.isArray(this.data) && this.data.length > 0;
    }

    override async processData(): Promise<void> {
        const { childrenKey, sizeKey, colorKey, fills, strokes, colorRange } = this;

        let index = 0;
        const getIndex = () => {
            index += 1;
            return index;
        };

        let maxDepth = 0;
        let minColor = Infinity;
        let maxColor = -Infinity;
        const colors: (number | undefined)[] = new Array((this.data?.length ?? 0) + 1).fill(undefined);

        const createNode = (datum: any, parent: HierarchyNode): HierarchyNode => {
            const index = getIndex();
            const depth = parent.depth != null ? parent.depth + 1 : 0;
            const children = childrenKey != null ? datum[childrenKey] : undefined;
            const isLeaf = children == null || children.length === 0;

            let size = sizeKey != null ? datum[sizeKey] : undefined;
            if (Number.isFinite(size)) {
                size = Math.max(size, 0);
            } else {
                size = isLeaf ? 1 : 0;
            }

            const sumSize = size;
            maxDepth = Math.max(maxDepth, depth);

            const color = colorKey != null ? datum[colorKey] : undefined;
            if (typeof color === 'number') {
                colors[index] = color;
                minColor = Math.min(minColor, color);
                maxColor = Math.max(maxColor, color);
            }

            return appendChildren(
                new HierarchyNode(this, index, datum, size, undefined, undefined, sumSize, depth, parent, []),
                children
            );
        };

        const appendChildren = (node: Mutable<HierarchyNode>, data: S[] | undefined): HierarchyNode => {
            data?.forEach((datum) => {
                const child = createNode(datum, node);
                node.children.push(child);
                node.sumSize += child.sumSize;
            });
            return node;
        };

        const rootNode = appendChildren(
            new HierarchyNode(this, 0, undefined, 0, undefined, undefined, 0, undefined, undefined, []),
            this.data
        );

        let colorScale: ColorScale | undefined;
        if (colorRange != null && Number.isFinite(minColor) && Number.isFinite(maxColor)) {
            colorScale = new ColorScale();
            colorScale.domain = [minColor, maxColor];
            colorScale.range = colorRange;
            colorScale.update();
        }

        rootNode.children.forEach((child, index) => {
            child.walk((node: Mutable<HierarchyNode>) => {
                let fill: string | undefined;

                const color = colors[node.index];
                if (color != null) {
                    fill = colorScale?.convert(color);
                }

                fill ??= fills?.[index % fills.length];

                node.fill = fill;
                // FIXME: If there's a color scale, the strokes won't make sense. For now, just hard-code this default
                node.stroke = colorScale == null ? strokes?.[index % strokes.length] : 'rgba(0, 0, 0, 0.2)';
            });
        });

        this.rootNode = rootNode;
        this.maxDepth = maxDepth;
    }

    getDatumIdFromData(node: HierarchyNode) {
        return `${node.index}`;
    }

    getDatumId(node: HierarchyNode) {
        return this.getDatumIdFromData(node);
    }
}
