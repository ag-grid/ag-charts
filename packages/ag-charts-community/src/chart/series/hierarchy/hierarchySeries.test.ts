import type { BBox } from '../../../scene/bbox';
import type { ChartAxisDirection } from '../../chartAxisDirection';
import type { ChartLegendDatum, ChartLegendType } from '../../legendDatum';
import type { SeriesNodeDataContext } from '../series';
import type { SeriesTooltip } from '../seriesTooltip';
import { type HierarchyNode, HierarchySeries } from './hierarchySeries';

class ExampleHierarchySeries extends HierarchySeries<any> {
    override tooltip: SeriesTooltip<any> = null!;

    override getSeriesDomain(_direction: ChartAxisDirection): any[] {
        throw new Error('Method not implemented.');
    }

    override createNodeData(): Promise<SeriesNodeDataContext<any, any>[]> {
        throw new Error('Method not implemented.');
    }

    override update(_opts: { seriesRect?: BBox | undefined }): Promise<void> {
        throw new Error('Method not implemented.');
    }

    override getTooltipHtml(_seriesDatum: any): string {
        throw new Error('Method not implemented.');
    }

    override getLegendData(_legendType: unknown): ChartLegendDatum<any>[] | ChartLegendDatum<ChartLegendType>[] {
        throw new Error('Method not implemented.');
    }
}

describe('hierarchySeries', () => {
    it('Creates a hierarchy', async () => {
        const series = new ExampleHierarchySeries(null!);
        series.sizeKey = 'size';
        series.data = [
            { size: 5, children: [{ size: 1 }, { size: 2 }, { size: 3 }] },
            {
                size: 5,
                children: [
                    { size: 1 },
                    { size: 2, children: [{ size: 4 }, { size: 5 }, { size: 6 }] },
                    { size: 3, children: [{ size: 7 }] },
                ],
            },
        ];
        await series.processData();

        const removeDatum = (node: HierarchyNode<any>) => {
            delete node.datum;
            node.children.forEach(removeDatum);
        };

        removeDatum(series.rootNode);

        expect(series.rootNode).toMatchSnapshot();
        expect(series.sumSize).toBe(5 + 1 + 2 + 3 + 5 + 1 + 2 + 4 + 5 + 6 + 3 + 7);
        expect(series.maxDepth).toBe(3);
    });

    it('Handles an empty dataset', async () => {
        const series = new ExampleHierarchySeries(null!);
        series.data = [];
        await series.processData();

        expect(series.rootNode).toEqual({
            index: 0,
            datum: undefined,
            size: undefined,
            color: undefined,
            children: [],
        });
        expect(series.sumSize).toBe(0);
        expect(series.maxDepth).toBe(0);
    });

    it('Walks tree in pre-order', async () => {
        const series = new ExampleHierarchySeries(null!);
        series.data = [
            {
                order: 1,
                children: [{ order: 2 }, { order: 3 }, { order: 4 }],
            },
            {
                order: 5,
                children: [
                    { order: 6 },
                    { order: 7, children: [{ order: 8 }, { order: 9 }, { order: 10 }] },
                    { order: 11, children: [{ order: 12 }] },
                ],
            },
        ];
        await series.processData();

        let index = 0;
        series.walk((node) => {
            expect(node.index).toBe(index);

            if (node.datum != null) {
                expect(node.datum.order).toBe(index);
            }

            index += 1;
        });
        expect(index).toBe(12 + 1);
    });
});
