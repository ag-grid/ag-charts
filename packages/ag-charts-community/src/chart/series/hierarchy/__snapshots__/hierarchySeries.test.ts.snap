// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HierarchySeries creates a hierarchy 1`] = `
HierarchyNode {
  "children": [
    HierarchyNode {
      "children": [
        HierarchyNode {
          "children": [],
          "depth": 1,
          "fill": "#c16068",
          "index": 2,
          "midPoint": {
            "x": 0,
            "y": 0,
          },
          "size": 1,
          "stroke": "#874349",
          "sumSize": 1,
        },
        HierarchyNode {
          "children": [],
          "depth": 1,
          "fill": "#c16068",
          "index": 3,
          "midPoint": {
            "x": 0,
            "y": 0,
          },
          "size": 2,
          "stroke": "#874349",
          "sumSize": 2,
        },
        HierarchyNode {
          "children": [],
          "depth": 1,
          "fill": "#c16068",
          "index": 4,
          "midPoint": {
            "x": 0,
            "y": 0,
          },
          "size": 3,
          "stroke": "#874349",
          "sumSize": 3,
        },
      ],
      "depth": 0,
      "fill": "#c16068",
      "index": 1,
      "midPoint": {
        "x": 0,
        "y": 0,
      },
      "size": 5,
      "stroke": "#874349",
      "sumSize": 11,
    },
    HierarchyNode {
      "children": [
        HierarchyNode {
          "children": [],
          "depth": 1,
          "fill": "#a2bf8a",
          "index": 6,
          "midPoint": {
            "x": 0,
            "y": 0,
          },
          "size": 1,
          "stroke": "#718661",
          "sumSize": 1,
        },
        HierarchyNode {
          "children": [
            HierarchyNode {
              "children": [],
              "depth": 2,
              "fill": "#a2bf8a",
              "index": 8,
              "midPoint": {
                "x": 0,
                "y": 0,
              },
              "size": 4,
              "stroke": "#718661",
              "sumSize": 4,
            },
            HierarchyNode {
              "children": [],
              "depth": 2,
              "fill": "#a2bf8a",
              "index": 9,
              "midPoint": {
                "x": 0,
                "y": 0,
              },
              "size": 5,
              "stroke": "#718661",
              "sumSize": 5,
            },
            HierarchyNode {
              "children": [],
              "depth": 2,
              "fill": "#a2bf8a",
              "index": 10,
              "midPoint": {
                "x": 0,
                "y": 0,
              },
              "size": 6,
              "stroke": "#718661",
              "sumSize": 6,
            },
          ],
          "depth": 1,
          "fill": "#a2bf8a",
          "index": 7,
          "midPoint": {
            "x": 0,
            "y": 0,
          },
          "size": 2,
          "stroke": "#718661",
          "sumSize": 17,
        },
        HierarchyNode {
          "children": [
            HierarchyNode {
              "children": [],
              "depth": 2,
              "fill": "#a2bf8a",
              "index": 12,
              "midPoint": {
                "x": 0,
                "y": 0,
              },
              "size": 7,
              "stroke": "#718661",
              "sumSize": 7,
            },
          ],
          "depth": 1,
          "fill": "#a2bf8a",
          "index": 11,
          "midPoint": {
            "x": 0,
            "y": 0,
          },
          "size": 3,
          "stroke": "#718661",
          "sumSize": 10,
        },
      ],
      "depth": 0,
      "fill": "#a2bf8a",
      "index": 5,
      "midPoint": {
        "x": 0,
        "y": 0,
      },
      "size": 5,
      "stroke": "#718661",
      "sumSize": 33,
    },
  ],
  "depth": undefined,
  "fill": undefined,
  "index": 0,
  "midPoint": {
    "x": 0,
    "y": 0,
  },
  "size": 0,
  "stroke": undefined,
  "sumSize": 44,
}
`;
